### PR TITLE
Add a new type of event for policy action notification. 

### LINKED ISSUE #9914: Optimized eBPF trace for forward action logging (datapath)
### Summary

The current datapath supports trace and capture notifications on drops as well as on various entry and exit points. In order to support the understanding of what policy rules are being utilized, a trace message is required which is submitted on first occurrence of a flow.

### Details

The current trace notifications are intended for troubleshooting and debugging. While they offer great visibility, their impact can be high. There is a minimal optimization in place or TCP based on TCP flags but a connection tracker based approach will help further reduce the overhead to gain visibility.

### Tasks

- [ ] Define metadata for trace message
- [ ] Add ability to emit trace message on `CT_NEW`
- [ ] Add datapath option to disable/enable decoupled from existing trace message flag to allow enabling this lower frequency message without requiring to enable any per packet notifications.
- [ ] Add `cilium monitor` parser

### PR DESCRIPTION
**Pull request by lzang**
A new event type for policy verdict is added in BPF. It can be turned on/off using endpoint build option PolicyVerdictNotification.  The event can be monitored with "cilium monitor -t policy-verdict"

Signed-off-by: Zang Li <zangli@google.com>
Fixes: #9914

```release-note
Add a new event type for policy verdicts
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9943)

### CODE REVIEW

File: bpf/lib/common.h
Comment (joestringer) [inline bpf/lib/common.h:262]: The current usage of these is a bit inconsistent, specifically as of review this is how these variables are used:

`POLICY_MATCH_L3` indicates l3-only match
`POLICY_MATCH_L4` indicates L3 + L4 match (known as l3-dependent l4 allow)
`POLICY_MATCH_L4_ALL` indicates L4-only match
`POLICY_MATCH_ALL` indicates allow-all match

To make this more clear, I would suggest something like:

* `POLICY_MATCH_L3_ONLY`
* `POLICY_MATCH_L3_L4`
* `POLICY_MATCH_L4_ONLY`
* `POLICY_MATCH_ALL`

File: bpf/lib/policy_log.h
Comment (joestringer) [inline bpf/lib/policy_log.h]: nit:
```suggestion
 * TODO: add DROP_NOTIFY build option
```

File: bpf/bpf_lxc.c
Comment (joestringer) [inline bpf/bpf_lxc.c]: It's probably worth noting here that if the verdict is > 0 then it is subject to L7 policy, and may or may not be allowed. I'd expect such cases to show up in the monitor output with "action: allow, type: none" with the current implementation.

Not sure how you want to notify about such traffic which is not yet allowed nor denied?

File: pkg/monitor/dissect.go
Comment (joestringer) [inline pkg/monitor/dissect.go]: Did you consider sharing more of this implementation with `GetConnectionSummary()`?

File: pkg/option/runtime_options.go
Comment (joestringer) [inline pkg/option/runtime_options.go]: I think that it would be useful to be more clear in general about these notifications that they are policy *verdict* notifications. They are notifications coming from the datapath about a decision being made. This would help to differentiate from potential other policy notifications like if we were to notify when changes to policy are applied to the datapath.

File: bpf/bpf_lxc.c
Comment (joestringer) [reply bpf/bpf_lxc.c]: To make the verdict handling more consistent across callers of `send_policy_notify4()`, it might be worth considering just handing the verdict down to that function for it to handle how to interpret the verdict.

**Comment by joestringer** [comment]
test-me-please

**Review by joestringer — CHANGES_REQUESTED** [review]
Thanks for the contribution!

I have a couple of concerns in this PR as it stands around the way that it portrays the events it is sending. Should be a fairly straightforward fix, but specifically:
* The wording tends to describe these events as "policy" events without much other qualification. Given that Cilium is responsible for implementing the policy both from a control plane and dataplane perspective, I think it would be good to make sure that these events are clearly communicated as policy verdict events, ie events triggered from a policy action taken upon traffic passing through. I could imagine having other kinds of policy events in future, and it'd be easier to make this point clear now.
* The specific representation of L3-only, L4-only, L3-dependent-L4, wildcard all, and L7 allows in the datapath seemed a bit inconsistent to me, for more details see the comments. We can discuss further below.

Although you have the sign-off in the PR description, the bot only recognizes it in the git commit messages which is where the history is properly stored. Please also sign-off your commit messages. If the maintainers-little-helper link does not describe this clearly enough, feel free to ask one of us in the #development channel on Slack.

I kicked off the CI, which you should be able to investigate later by clicking the link after "Cilium-Ginkgo-Tests" in the checks box at the bottom of this PR. Note that if you re-push, this will disappear. We have a few flakes known on master at the moment, so again feel free to ping us on Slack to help interpret the Jenkins output if it's not green.

This was just my initial drive-by review, I didn't think too deeply about the format just yet.

**Comment by lzang** [comment]
@joestringer I sent this PR so that we can start the discussion, and I should have marked it as WIP :-) Thank you for the good points you made in the comments! I will update the patches. 
Question: if a traffic is redirected, how to know its final action? 
Also the Cilium-Ginkgo-Tests shows failed, but the "Details" link returns 404.

**Comment by joestringer** [comment]
> Question: if a traffic is redirected, how to know its final action?

That is up to the L7 proxy implementation, be it the DNS proxy inside `cilium-agent` or the other protocol handlers implemented in envoy. I believe that cilium already sends notifications for these on a per-request level rather than per connection (well, because inherently the L7 filtering is applying to more specific individual requests in the connection compared to the L3/L4 filtering which can only determine how to handle entire connections).

> Also the Cilium-Ginkgo-Tests shows failed, but the "Details" link returns 404.
The CI will clean these up after some time, usually I thought it takes longer than this. However right now we're hitting some infrastructure issues there so it's probably better for us to focus on the overall design and objectives rather than the individual details of this particular implementation and whether it passes CI.

File: bpf/bpf_lxc.c
Comment (lzang) [reply bpf/bpf_lxc.c]: With the current implementation, type will not be none as the redirect is the result of policy being matched. I am thinking to change it to "action: redirect". Is redirected port needed to be included  in the message for this case? Would that port number change across different L7 rules?

Comment (joestringer) [reply bpf/bpf_lxc.c]: Ah, you're right. I misread the first time. "action: redirect" seems reasonable to me. Port could potentially be useful; current implementation uses one unique port per protocol.

File: pkg/monitor/dissect.go
Comment (lzang) [reply pkg/monitor/dissect.go]: Removed the function for now.

File: pkg/option/runtime_options.go
Comment (lzang) [reply pkg/option/runtime_options.go]: Updated.

File: bpf/bpf_lxc.c
Comment (lzang) [reply bpf/bpf_lxc.c]: Updated and include the redirect port.

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: pkg/monitor/api/types.go
Comment (joestringer) [inline pkg/monitor/api/types.go]: The comment above this one uses a different name from the variable below.

File: pkg/monitor/datapath_policy.go
Comment (joestringer) [inline pkg/monitor/datapath_policy.go:66]: Please add a reference here to the C structure that this corresponds to. Also, I think that we should be able to use `pkg/datapath/alignchecker` to validate that the C and Go structures match. See also the usage for `trace_notify` structure.

File: bpf/lib/policy_log.h
Comment (joestringer) [inline bpf/lib/policy_log.h]: Did you consider folding the flags in the next u8 into the same `match_type` field here? Seems like we could save a u32 and shuffle the fields closer together.

Even if we do this and leave the u32 padding at the end of the structure I think it would help future extension by keeping a larger set of unused bits together.

Do you think that we will likely extend this structure to have more information in future?

File: bpf/lib/policy_log.h
Comment (joestringer) [inline bpf/lib/policy_log.h:40]: Does a 0 bit here imply IPv4? What if we drop other, non-IP traffic?

**Review by joestringer — CHANGES_REQUESTED** [review]

File: bpf/lib/policy_log.h
Comment (lzang) [reply bpf/lib/policy_log.h]: Sure, I can make that change. I would like to get some suggestions on if more information should be included in this structure. I think it depends on what fields are useful for the quick rule association in the user plane. match_type was suggested by @tgraf

File: bpf/lib/policy_log.h
Comment (lzang) [reply bpf/lib/policy_log.h:40]: This log is for policy verdict. Does Cilium support policy for non-IP traffic? 
trace_notify also just used 1 bit for ipv6 indication.
If there is a possibility to log non-ip traffic, we can extend it into 2 bits and could you please suggest a field name? Thanks.

Comment (joestringer) [reply bpf/lib/policy_log.h:40]: If it's consistent with `trace_notify` then maybe this is enough for now. I was thinking that when we enable policy, we only support allows for IPv4/IPv6 so we could potentially drop other traffic, but you're right, we won't generate the monitor message in that case.

I wouldn't worry about it, we can always add other bits later.

File: bpf/lib/policy_log.h
Comment (joestringer) [reply bpf/lib/policy_log.h]: I could imagine a path where userspace `cilium-agent` creates identifiers for each policymap entry and pushes these down into the policymap, then when doing policy verdict tracing with these messages, those identifiers are passed back up in the monitor message.

If these are allocated by the agent, then 16 bits (65K ids) seems like plenty of uniqueness for an individual policy rule on a single node.
If these end up being centrally managed, then I could imagine desire for this to be a UUID.

Comment (lzang) [reply bpf/lib/policy_log.h]: I considered this way of implementation, but it is pretty complicated because a traffic can match multiple policies and it incurs extra effort to manage the identifier in this case. I feel it is probably not the direction Cilium wants to pursue. On the other hand, how difficult it is to add fields later? or we leave some padding there for future use ?

Comment (joestringer) [reply bpf/lib/policy_log.h]: I would imagine the policy multiplexing would occur in userspace to avoid creating a list of identifiers in the dataplane. But I agree, this is complicated.

In the absence of other feedback regarding upcoming features, I would be inclined to just round it to the nearest allocation boundary and leave it at this.

Comment (lzang) [reply bpf/lib/policy_log.h]: Updated the patch after the offline discussion. Added proto and port fields to the event and rearranged padding.

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please
The DirectRouting test failed for k8s-1.11 but passed for k8s 1.17.
Link: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17010/
Rerun the test to see whether this is flake or it is related to my change.

**Comment by joestringer** [comment]
@lzang the current implementation of `test-me-please` requires that to to be the entire content of the GitHub comment, so you would need to post two messages to provide the above info.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Latest failure seems to be provisioning issue: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17023/

**Comment by joestringer** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Review by joestringer — CHANGES_REQUESTED** [review]
Looks good, just need the IPv6 functions corresponding to the IPv4 versions here.

**Comment by lzang** [comment]
test-me-please

**Comment by aanm** [comment]
@nebril please white-list @lzang to trigger the CI

**Comment by aanm** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by lzang** [comment]
test-me-please

**Comment by aanm** [comment]
@joestringer changes LGTM, PTAL again. I believe the IPv6 is added as well

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by tgraf** [comment]
CI failed due to what appear to be infrastructure or provisioning related failures.

**Comment by tgraf** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by lzang** [comment]
test-me-please

**Comment by joestringer** [comment]
@lzang did you address these already from the recent CI run?

```
10:14:19  In file included from bpf_lxc.c:52:
10:14:19  /go/src/github.com/cilium/cilium/bpf/lib/policy_log.h:50:2: error: use of undeclared identifier 'uint64_t'
10:14:19          uint64_t skb_len = (uint64_t)skb->len, cap_len = min((uint64_t)TRACE_PAYLOAD_LEN, (uint64_t)skb_len);
10:14:19          ^
10:14:19  /go/src/github.com/cilium/cilium/bpf/lib/policy_log.h:51:2: error: use of undeclared identifier 'uint32_t'
10:14:19          uint32_t hash = get_hash_recalc(skb);
10:14:19          ^
10:14:19  /go/src/github.com/cilium/cilium/bpf/lib/policy_log.h:55:11: error: use of undeclared identifier 'hash'
10:14:19                  .hash = hash,
10:14:19                          ^
10:14:19  /go/src/github.com/cilium/cilium/bpf/lib/policy_log.h:56:15: error: use of undeclared identifier 'skb_len'
10:14:19                  .len_orig = skb_len,
10:14:19                              ^
10:14:19  /go/src/github.com/cilium/cilium/bpf/lib/policy_log.h:57:14: error: use of undeclared identifier 'cap_len'
10:14:19                  .len_cap = cap_len,
10:14:19                             ^
10:14:19  /go/src/github.com/cilium/cilium/bpf/lib/policy_log.h:71:6: error: use of undeclared identifier 'cap_len'
10:14:19                           (cap_len << 32) | BPF_F_CURRENT_CPU,
10:14:19                            ^
10:14:19  6 errors generated.
```

**Comment by lzang** [comment]
Yes, just noticed that and updated the patch.

**Comment by lzang** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Pull request merged**
